### ISSUE TITLE
Updates to Cookie Banner

### ISSUE THREAD

**Issue by edthedev** [app-illinois/Design-Resources#133]
## Context

Work on cookie banner continues from #128.

## Tasks

- [x] get the page merged into `develop` and live at https://app-illinois.github.io/Design-Resources/Cookie-Banner.html
- [x] email [accessibility@illinois.edu](mailto:accessibility@illinois.edu) to request accessibility feedback (include the live demo link at https://app-illinois.github.io/Design-Resources/Cookie-Banner.html )
- [x] #136
- [x] update with any feedback from accessibility
- [x] allow the slide-out on narrow mobile views to expand to use the full screen width - add a media query for mobile

## Feedback from April 2 meeting

- [x] #137
- [x] confirm update to v3 resulted in `hover` and `focus` button states being visually distinguishable from each-other
- [x] add some space above `About Cookies` inside the main banner.
- [x] change the cursor pointer to indicate that the slide-down sections are clickable in the cookie details collapsible headers.
- [x] add a close button to the bottom of the slide-out, so that users don't have to back-track to the `X` to close it.

## Stretch Goal Tasks (or lovability updates)

- [x] #138
- [x] stretch: add CSS that makes the Cookie Banner not visible in print layouts.
- [x] #139
- [x] #140

**Comment by mpitcel** [comment]
Can we bundle this as a JS Script for easier replacement?

We can request help if we cannot do this type of update.

**Comment by mpitcel** [comment]
Other items that may be good to change:

- Pointer change on Accordions 
- More padding on the Banner Header 
- Check on Button behavior (on hover, etc.) 
- Update to V3 from V2 may change the Button as desired

**Comment by edthedev** [comment]
We've captured this elsewhere, but it's good to include it here, as well. The next release is roughly a one-for-one replacement (with accessibility improvements) of our current vendor capabilities, and so won't include any controls to accept/reject cookies. 

We (particularly our privacy team) aren't satisfied with that, long term. We will continue to work on updating this cookie banner to enable stronger privacy practices.

**Comment by mpitcel** [comment]
> stretch: avoid zombie banner behavior of returning each session when the page reloads - likely limited to current subdomain, but ideally acknowledgement lasts as long and wide as browser security settings allow

This may be actionable by utilizing the Session Storage:

- https://developer.mozilla.org/en-US/docs/Web/API/Window/sessionStorage 
- https://www.w3schools.com/jsref/prop_win_sessionstorage.asp

**Comment by mpitcel** [comment]
> update that upstream CSS is on v3 instead of v2. Note that v2 and v3 can be used side-by-side, so it's okay that upstream components we use are still on v2.

Per the guidance at https://app-illinois.github.io/Design-Resources/ , it appears that this is achieved by updating the items in `<head>` with most of those in the `template`.

**Comment by mpitcel** [comment]
Found a neat way to check Accessibility states of elements in Edge: https://learn.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/accessibility/test-inspect-states

**Comment by edthedev** [comment]
## Cookie Banner Accessibility Review

Extracted details from our accessibility review are included below. I have a copy of the full PDF and am happy to share, but I've copied critical details here to facilitate code updates needed to address issues.

## Cookie Banner Accessibility Review Summary

This review focused almost entirely upon the Cookie Notice rather than the base page. There are
no Showstoppers and only a few P1 issues. The most serious violations are in the Mobile
environment. For the most part the application is in good shape

## Priority 1 Accessibility feedback:

- [x] P1 – The link, “University of Illinois System Cookie Policy” inside the About Cookies dialog
window opens a new window without alerting the user. It’s a best-practice to let users know
this is going to happen so that they can better navigate the site. (Reference WCAG 2.1 -
2.4.4 Link Purpose (A) - https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-
mechanisms-refs.html, G201: Giving users advanced warning when opening a new window -
https://www.w3.org/TR/WCAG20-TECHS/G201.html)
- [x] When the GitHub page initially loads, focus does not start with the Cookie Notice
Banner but rather begins at the top of the page and requires one to swipe through the
entire page before moving into the Cookie Notice. Once the user reaches the Cookie Notice,
the Close X takes the initial focus and then one moves to the H2 Cookie Notice title. The
user will not initially know they’ve reached the Cookie Notice.
- [x] a. Additionally, the Cookie Notice covers the bottom 50% of the screen, however as one
swipes through the window, focus continues on the web page, going under the Cookie
Notice and is no longer visible. The expected behavior is for the Cookie Notice to capture
focus initially once the page loads and not release it until the user acknowledges either
their acceptance or rejection of Cookies on the site, at which time the Cookie Notice
disappears. (Reference WCAG 2.1 - 2.4.3 Focus Order (A) -
https://www.w3.org/TR/UNDERSTANDING-WCAG20/navigation-mechanisms-focus-
order.html)
- [x] On the Mobile screen, when the About Cookies
dialog window opens, if the screen is in Portrait view, it
is squeezed into the left side of the screen and becomes
illegible, see Figure 1. Users should not have to adjust
screen orientation in order to see the content.
(Reference WCAG 2.1 - 1.3.4 Orientation (AA) -
https://www.w3.org/WAI/WCAG21/Understanding/
orientation.html)

Priority 2 Accessibility feedback:

- [x] When the page initially opens, shouldn’t focus go directly to the Cookie Banner bar? As
it is, tab focus order places the Cookie Banner at tabs 5 through 8. Blind users may not be
aware the Cookie banner is there.
- [ ] The initial description says, “By continuing to browse without changing your browser
settings to block or delete Cookies…” which assumes the user knows how to make this
change in their browser. Isn’t the option to block or delete cookies something that should
be offered within the Cookie Banner? This banner assumes one is fine with active cookies
since it doesn’t give the user the option to choose anything else. Even the About Cookies
dialog does not offer any option to control cookies. (ED notes: This feedback will be addressed in a follow-up update, after the first release. The first release is scoped to only replace the current functionality of the purchased cookie banner. But feedback reminds us that this isn't an acceptable place to stop with our cookie banner solution. We intend to address this in future updates.)
- [x] The following findings are related to the “About Cookies” button:

    a. When one activates the “About Cookies” button, a dialogue window opens up but
nothing is announced alerting the user that something has happened.
    b. After the “Illinois” dialog window opens, when the X button is focused on, it announces,
“close slideover, button, Illinois, dialog”. The inclusion of “slideover” in this
announcement is unnecessary since the user isn’t going to know that the dialog window
is a slideover element.
    c. The “Illinois” H2 title for the About Cookies dialog window doesn’t describe what this
window is for. A more descriptive title would be beneficial.

- [ ] (Not related to the Cookie Notice) On an Android
phone, when one dismisses the Cookie Notice and
swipes off the last link in the main part of the page,
“Link to this standard in GitHub”, the next two swipes
are silent and have no focus, nor do they activate
anything. When using VoiceOver on an iPhone, swiping
off the last link goes directly to the U of I icon link as
expected. (ED notes: While this isn't directly part of the cookie banner, we also want our demo pages to be accessible, and so this is in scope for our accessibility update.)

ED: While none of the items above were identified as show-stoppers, I am asking the SecDev team to look at each item above for our next release, except for cookie management, which needs to be scoped for a follow-up release.

**Comment by mpitcel** [comment]
I am currently working to make the Cookie Popup behave such that it must be acknowledged before the User can interact with the page.

**Comment by edthedev** [comment]
> I am currently working to make the Cookie Popup behave such that it must be acknowledged before the User can interact with the page.

Per our discussion, we do want the banner to take initial focus, to avoid accessibility issues, but people don't expect to be forced to interact with the cookie banner, and may want to jump directly to some other part of the page. Per prior discussions, we should not make it modal, but the cookie banner should stay around until dismissed by user interaction.

Let's leave the relevant code in as disabled comments, for now, so teams have the option to enforce it.

**Comment by edthedev** [comment]
For V3 - https://builder3.toolkit.illinois.edu/getting_started/index.html
